File: dali/operators/image/resize/experimental/resize.cc
// Copyright (c) 2024-2025, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#define DALI_RESIZE_BASE_CC

#include "dali/operators/image/resize/experimental/resize.h"
#include <cassert>
#include "dali/pipeline/data/views.h"

namespace dali {

DALI_SCHEMA(experimental__Resize)
  .DocStr(R"code(Resize images.)code")
  .NumInput(1)
  .NumOutput(1)
  .AdditionalOutputsFn([](const OpSpec& spec) {
    return static_cast<int>(spec.GetArgument<bool>("save_attrs"));
  })
  .InputLayout(0, {"HWC",  "FHWC",  "CHW",  "FCHW",  "CFHW" ,
                   "DHWC", "FDHWC", "CDHW", "FCDHW", "CFDHW"  })
  .AddOptionalArg("save_attrs",
      R"code(Save reshape attributes for testing.)code", false)
  .AddOptionalArg<DALIImageType>("image_type", "Image type", nullptr)
  .DeprecateArg("image_type", "0.25")
  .SupportVolumetric()
  .AllowSequences()
  .AddParent("ResizeAttr")
  .AddParent("ResamplingFilterAttr");

CvCudaResize::CvCudaResize(const OpSpec &spec)
    : StatelessOperator<GPUBackend>(spec)
    , ResizeBase<GPUBackend>(spec) {
  save_attrs_ = this->spec_.HasArgument("save_attrs");
  InitializeBackend();
}

void CvCudaResize::InitializeBackend() {
  InitializeGPU(spec_.GetArgument<int>("minibatch_size"),
                spec_.GetArgument<int64_t>("temp_buffer_hint"));
}

void CvCudaResize::RunImpl(Workspace &ws) {
  const auto &input = ws.Input<GPUBackend>(0);
  auto &output = ws.Output<GPUBackend>(0);

  RunResize(ws, output, input);
  output.SetLayout(input.GetLayout());

  if (save_attrs_) {
    auto &attr_out = ws.Output<GPUBackend>(1);
    const auto &attr_shape = attr_out.shape();
    assert(attr_shape.num_samples() == input.shape().num_samples() &&
           attr_shape.sample_dim() == 1 &&
           is_uniform(attr_shape) &&
           attr_shape[0][0] == NumSpatialDims());

    if (!attr_staging_.has_data())
      attr_staging_.set_pinned(true);
    attr_staging_.Resize(attr_out.shape(), DALI_INT32);
    auto attr_view = view<int, 1>(attr_staging_);
    SaveAttrs(attr_view, input.shape());
    attr_out.Copy(attr_staging_, ws.stream());
  }
}

DALI_REGISTER_OPERATOR(experimental__Resize, CvCudaResize, GPU);

}  // namespace dali

File: dali/python/nvidia/dali/experimental/torchvision/v2/functional/resize.py
# Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import Optional, List, Literal
import nvidia.dali.experimental.dynamic as ndd
from torchvision.transforms import InterpolationMode

import torch
from PIL import Image

from ..operator import adjust_input, get_HWC_from_layout_dynamic  # noqa: E402
from ..resize import Resize  # noqa: E402


@adjust_input
def resize(
    inpt: Image.Image | torch.Tensor,
    size: List[int],
    interpolation: InterpolationMode = InterpolationMode.BILINEAR,
    max_size: Optional[int] = None,
    antialias: Optional[bool] = True,
    device: Literal["cpu", "gpu"] = "cpu",
) -> Image.Image | torch.Tensor:
    """
    Please refer to the ``Resize`` operator for more details.
    """
    Resize.verify_args(
        size=size, max_size=max_size, interpolation=interpolation, antialias=antialias
    )

    size_normalized = Resize.infer_effective_size(size)
    interpolation = Resize.interpolation_modes[interpolation]

    original_h, original_w, _ = get_HWC_from_layout_dynamic(inpt)

    target_h, target_w = Resize.calculate_target_size_dynamic_mode(
        (original_h, original_w), size_normalized, max_size
    )

    return ndd.resize(
        inpt,
        device=device,
        size=(target_h, target_w),
        interp_type=interpolation,
        antialias=antialias,
    )
